### PR TITLE
Thruster plugin: add tests and velocity control

### PR DESCRIPTION
**Pull request by chapulina**
## Summary

The `Thruster` plugin was added in #749 without tests. This PR adds tests, and while at it, I also:

* Documentation
    * Documented `examples/worlds/auv_controls.sdf`
    * Documented missing XML fields in `Thruster.hh`
    * Updated variables and docs that mentioned RPM, because the plugin uses SI for angular velocity (rad / s)
* XML params
    * Some parameters were listed as "required", but also offered a default value. I made the implementation consistent with documentation by making parameters that have default values be optional.
    * Fail loading the plugin if the user gives a joint name that doesn't exist
    * Default the namespace to the model name instead of empty; empty is weird because the topic becomes `/model/joint/...` - this can potentially break users that are currently leaving the namespace empty.
* Implementation
    * Compute the desired propeller angular velocity only when the desired thrust changes
    * I think the `cmd_pos` topic is very confusing for this plugin, because the commands are forces in Newtons. So I'm proposing changing the topic to `cmd_thrust`. I didn't remove `cmd_pos` for backwards compatibility, but I think we should tick-tock it out.
    * Take the `<joint><pose>` into account
* Velocity control
    * Ported the alternative joint velocity control from downstream (https://github.com/osrf/lrauv/commit/c20b44c3b28393ea6929ab22073009dbb7a94999).
    * The motivation for this feature was that the PID was accumulating errors that made the propellers drift over time
    * While writing tests, it became clear that using `JointVelocityCmd` makes the vehicle much more unstable than directly applying the wrench to the link. So there's a trade-off.

## Test it

Follow the instructions in `auv_controls.sdf`.

Here are some handy commands to run the test worlds too:

```
ign gazebo -r -v 4 src/ign-gazebo/test/worlds/thruster_pid.sdf
ign topic -t /model/sub/joint/propeller_joint/cmd_thrust  -m ignition.msgs.Double -p 'data: 0.3'
```

```
ign gazebo -r -v 4 src/ign-gazebo/test/worlds/thruster_vel_cmd.sdf
ign topic -t /model/custom/joint/propeller_joint/cmd_thrust  -m ignition.msgs.Double -p 'data: 0.3'
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Review by arjo129 — APPROVED** [review]
We should be OK to merge this as most of the issues are experienced at low velocities and I suspect due to issues with the hydrostatics.

**Comment by matosinho** [comment]
I see that @chapulina added TODO to configure max and min thrust in the sdf. This means that it will be done in another PR? I think that would be a good addition.

**Comment by chapulina** [comment]
> I see that @chapulina added TODO to configure max and min thrust in the sdf. This means that it will be done in another PR? I think that would be a good addition.

I wasn't planning on tackling it anytime soon, but I agree it would be a nice addition

File: test/integration/thruster.cc [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      EXPECT_NE(propellerEntity, kNullEntity);
+
+      propeller = Link(propellerEntity);
+      propeller.EnableVelocityChecks(_ecm);
+    }).
+  OnPostUpdate([&](const gazebo::UpdateInfo &_info,
+                            const gazebo::EntityComponentManager &_ecm)
+    {
+      dt = std::chrono::duration_cast<std::chrono::nanoseconds>(
+          _info.dt).count() * 1e-9;
```
Comment (arjo129) [inline test/integration/thruster.cc:93]: Knit: could be done with `std::duration_cast<std::chrono::seconds, double>(_info.dt).count()` thus avoiding the use of 1e-9.

File: test/integration/thruster.cc [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    // The joint velocity command adds some roll to the body which the PID
+    // wrench doesn't
+    if (_namespace == "custom")
+      EXPECT_NEAR(0.0, pose.Rot().Roll(), 0.1);
+    else
+      EXPECT_NEAR(0.0, pose.Rot().Roll(), _baseTol);
+  }
+
+  // omega = sqrt(thrust /
```
Comment (arjo129) [inline test/integration/thruster.cc:179]: Should we add the book reference in the test for clarity?

**Review by arjo129 — APPROVED** [review]
Thanks for doing this! The tests and up-streaming are looking good.

File: test/integration/thruster.cc (hunk 1 above)
Comment (chapulina) [reply test/integration/thruster.cc]: Good call 9a0642b

File: test/integration/thruster.cc (hunk 2 above)
Comment (chapulina) [reply test/integration/thruster.cc:179]: 9a0642b

**Comment by iche033** [comment]
> I see that @chapulina added TODO to configure max and min thrust in the sdf. This means that it will be done in another PR? I think that would be a good addition.

here's a pull request for this: https://github.com/ignitionrobotics/ign-gazebo/pull/1318

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**